lib/rack/utils.rb

```diff
@@ -480,6 +480,12 @@ def initialize(hash={})
         hash.each { |k, v| self[k] = v }
       end
 
+      # on dup/clone, we need to duplicate @names hash
+      def initialize_copy(other)
+        super
+        @names = @names.dup
```

Comment:
We should be getting the names off the `other` object.  Is there a way to get them?  If not, can you add a protected method?